CleanFeed: Add option to treat specified tags as mature

enchanted-sword opened the pull request:
### Description
Resolves [#1178](https://github.com/AprilSylph/XKit-Rewritten/issues/1178)
Adds an option to treat specified tags as mature in the same way specified blogs can be treated as mature

### Testing steps
Add a tag to the text field in the CleanFeed options and view a post with said tag

marcustyphoon commented:
I don't think we can rename the `localFlagging` preference parameter without ensuring that the user's current preference is migrated by making use of the "inherit" preference key, as seen in the last few entries of `no_recommended.json` and in `hide_mini_follow` in `tweaks.json`.

enchanted-sword commented:
> I don't think we can rename the `localFlagging` preference parameter without ensuring that the user's current preference is migrated by making use of the "inherit" preference key, as seen in the last few entries of `no_recommended.json` and in `hide_mini_follow` in `tweaks.json`.

Ah, makes sense, I'll set it to inherit the value of `cleanfeed.preferences.locaFlagging`.

marcustyphoon commented on src/scripts/cleanfeed.js:
```diff
@@ -10,33 +10,36 @@ const styleElement = buildStyle();
   const reblogs = postElement.querySelectorAll(reblogSelector);
   trail.forEach((trailItem, i) => {
-    if (trailItem.blog?.isAdult || localFlaggedBlogs.includes(trailItem.blog?.name)) {
+    if (trailItem.blog?.isAdult || localFlaggedBlogs.includes(trailItem.blog?.name || localFlaggedTags.some(t => trailItem.tags?.includes(t)))) {
```
Does this line function? That I know of, trail items don't have tag information.

enchanted-sword replied:
Whoops, I don't know why I added that. With the amount of work I put into my theme you'd think I would remember how NPF works 🤦‍♀️

marcustyphoon commented:
Any opinions on, sort of related to #982, the idea of allowing both `a tag` and `#a tag` in the input box by stripping any leading hash characters when processing the preference input?

enchanted-sword commented:
I did think about that, and my reasoning for not adding it was that having the default values `nsfw, nsft` with no hash prefix would indicate that the values don't need to have a hash prefix, but I suppose it wouldn't hurt to include it anyways. In the same vein, it might be a good idea to set the preference value strings to lowercase.

marcustyphoon commented on src/scripts/cleanfeed.js:
```diff
@@ -35,8 +40,9 @@ const processPosts = postElements => filterPostElements(postElements).forEach(as
-  localFlaggedBlogs = localFlagging.split(',').map(username => username.trim());
+  ({ blockingMode, localBlogFlagging, localTagFlagging } = await getPreferences('cleanfeed'));
+  localFlaggedBlogs = localBlogFlagging.toLowerCase().split(',').map(username => username.trim());
+  localFlaggedTags = localTagFlagging.toLowerCase().split(',').map(tag => tag.replace(/#/gi, '').trim());
```
In a simple case like this you can also use `.replaceAll('#', '')`; this may not be true in all codebases, but our browser support list is more than new enough! (https://caniuse.com/mdn-javascript_builtins_string_replaceall)

marcustyphoon reviewed (dismissed):
Functionality looks good to me! Not sure if anyone has opinions about the default contents of the tags field; I could imagine having one but I can't think of anything.

marcustyphoon commented:
Did that commit change anything, out of curiosity? `localBlogFlagging.split(',').toLowerCase()` would have been a bug, calling toLowerCase on an array, but `localBlogFlagging.toLowerCase().split(',')` seems fine.

(Not that it matters if it didn't; whichever way feels more clear is better!)

enchanted-sword commented:
> Did that commit change anything, out of curiosity? `localBlogFlagging.split(',').toLowerCase()` would have been a bug, calling toLowerCase on an array, but `localBlogFlagging.toLowerCase().split(',')` seems fine.

I think I spent too long looking at stylesheets yesterday lol, for some reason I looked at it again and thought I had tried to use `toLowerCase()` on an array so I changed it After this gets merged I was planning on working on #1188 so I can always just standardize it then.

marcustyphoon commented:
Yeah, to be fair, the fact that the previous way made us both have to think about (I started thinking about implicit type coercion) s a good sign that the new way is a clarity improvement :D

Thanks for contributing!
